### PR TITLE
Implement storage backends using RocksDB and SQLite

### PR DESCRIPTION
**Pull request by zah**

### CODE REVIEW

File: nimbus/db/storage_types.nim [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    # The first byte stores the key type. The rest are key-specific values
+    data: array[33, byte]
+    usedBytes: uint8
+
+  StorageError* = object of Exception
+
+proc genericHashKey*(h: Hash256): DbKey {.inline.} =
+  result.data[0] = byte ord(genericHash)
+  result.data[1 .. 32] = h.data
+  result.usedBytes = uint8 32
```
Comment (yglukhov) [inline nimbus/db/storage_types.nim]: `usedBytes` should be 33?

File: nimbus/db/storage_types.nim [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+proc blockHashToScoreKey*(h: Hash256): DbKey {.inline.} =
+  result.data[0] = byte ord(blockHashToScore)
+  result.data[1 .. 32] = h.data
+  result.usedBytes = uint8 32
```
Comment (yglukhov) [inline nimbus/db/storage_types.nim]: `usedBytes` should be 33?

File: nimbus/db/storage_types.nim [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+proc transactionHashToBlockKey*(h: Hash256): DbKey {.inline.} =
+  result.data[0] = byte ord(transactionHashToBlock)
+  result.data[1 .. 32] = h.data
+  result.usedBytes = uint8 32
```
Comment (yglukhov) [inline nimbus/db/storage_types.nim]: `usedBytes` should be 33?

File: nimbus/db/storage_types.nim [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+proc blockNumberToHashKey*(u: BlockNumber): DbKey {.inline.} =
+  result.data[0] = byte ord(blockNumberToHash)
+  assert sizeof(u) <= 32
+  copyMem(addr result.data[1], unsafeAddr u, sizeof(u))
+  result.usedBytes = uint8 sizeof(u)
```
Comment (yglukhov) [inline nimbus/db/storage_types.nim]: `usedBytes` should be sizeof(u) + 1?

File: nimbus/db/storage_types.nim [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+proc canonicalHeadHashKey*(): DbKey {.inline.} =
+  result.data[0] = byte ord(canonicalHeadHash)
+  result.usedBytes = 32
```
Comment (yglukhov) [inline nimbus/db/storage_types.nim]: `usedBytes` should be 1?

File: nimbus/db/storage_types.nim [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+const hashHolderKinds = {genericHash, blockHashToScore, transactionHashToBlock}
+
+template toOpenArray*(k: DbKey): openarray[byte] =
+  k.data.toOpenArray(0, int k.usedBytes)
+
+proc hash*(k: DbKey): Hash =
+  result = hash(k.toOpenArray)
+
+proc `==`*(a, b: DbKey): bool {.inline.} =
+  equalMem(unsafeAddr a, unsafeAddr b, sizeof(a))
```
Comment (yglukhov) [inline nimbus/db/storage_types.nim]: `usedBytes` instead of `sizeof(a)`?

**Comment by zah** [comment]
@yglukhov, my idea so far is that the back-end will depend only on the target platform and we won't provide a run-time option for selecting it, only a compile-time option.

The mobile builds will use SQLite and the server builds will use RocksDB.

**Comment by zah** [comment]
Also, polymorphism should be supported with VTable types in the long-run. The Trie repo currently includes a poor's implementation of a "StorageBackend" VTable type:

https://github.com/status-im/nim-eth-trie/blob/master/eth_trie/types.nim#L23

File: nimbus/db/storage_types.nim [hunk 7]
```diff
@@ -12,45 +12,56 @@ type
...
   result.data[0] = byte ord(blockNumberToHash)
   assert sizeof(u) <= 32
   copyMem(addr result.data[1], unsafeAddr u, sizeof(u))
-  result.usedBytes = uint8 sizeof(u)
+  result.dataEndPos = uint8 sizeof(u)
 
 proc canonicalHeadHashKey*(): DbKey {.inline.} =
   result.data[0] = byte ord(canonicalHeadHash)
-  result.usedBytes = 32
+  result.dataEndPos = 32
```
Comment (yglukhov) [inline nimbus/db/storage_types.nim]: Still this should be 1 :)

**Pull request merged**